base/strings/string.jl

```diff
@@ -1,7 +1,16 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+struct StringIndexError <: Exception
+    string::AbstractString
+    index::Integer
+end
+@noinline string_index_err(s::AbstractString, i::Integer) =
+    throw(StringIndexError(s, Int(i)))
+
 const ByteArray = Union{Vector{UInt8},Vector{Int8}}
 
+@inline between(b::T, lo::T, hi::T) where {T<:Integer} = (lo ≤ b) & (b ≤ hi)
```

Comment:
Isn't this just `lo ≤ b ≤ hi`? If the problem is the difference between `&&` and `&`, I think the compiler avoids branching in simple cases like this anyway.

Reply:
I was hoping so but the code gen was considerably different here. I would love to replace all uses of this with `lo ≤ b ≤ hi`.

Reply:
Verified, these do generate different code:
```jl
julia> between1(x, lo, hi) = lo <= x <= hi
between1 (generic function with 1 method)

julia> between2(x, lo, hi) = (lo <= x) & (x <= hi)
between2 (generic function with 1 method)

julia> @code_native between1(2, 1, 10)
	.section	__TEXT,__text,regular,pure_instructions
; Function between1 {
; Location: REPL[78]:1
	cmpq	%rdi, %rsi
	jle	L8
	xorl	%eax, %eax
	retq
L8:
	cmpq	%rdx, %rdi
	setle	%al
	retq
	nop
;}

julia> @code_native between2(2, 1, 10)
	.section	__TEXT,__text,regular,pure_instructions
; Function between2 {
; Location: REPL[79]:1
	cmpq	%rdi, %rsi
	setle	%cl
	cmpq	%rdx, %rdi
	setle	%al
	andb	%cl, %al
	retq
	nop
;}
```
The performance difference seems to be negligible here, but I worry that in the context of loops and other complex code having branches will make it much harder for the compiler to optimize the surrounding code.